crates/sui/src/sui_move/sui-natives.bpl

```diff
@@ -5,11 +5,90 @@
 // Native address
 
 
-procedure {:inline 1} $2_address_from_bytes(bytes: Vec (int)) returns (res: int);
-
-procedure {:inline 1} $2_address_to_u256(addr: int) returns (res: int);
-
-procedure {:inline 1} $2_address_from_u256(num: int) returns (res: int);
+const $MAX_ADDRESS: int;
+axiom $MAX_ADDRESS == 1461501637330902918203684832716283019655932542975;
+
+const $ADRESS_LENGTH: int;
+axiom $ADRESS_LENGTH == 20;
+
+// helper function for converting vector to address.
+function $2_peel_vector_to_address(bytes: Vec (int), len: int): int
+{
+    if len > 0 then 256 * $2_peel_vector_to_address(bytes, len-1) + ReadVec(bytes, len-1)
+    else 0
+}
```

Comment:
This function is defined both
- imperatively, as a recursive function AND
- axiomatically, via the following `axiom`s,

This seems a bit weird to me.

I would suggest to remove the function body here and leave `$2_peel_vector_to_address` an uninterpreted function that is defined by the following `axiom`s. In this way, it is oblivious to the actual implementation of the address conversion.

Reply:
I initially attempted to define the function using only axioms as you recommended, but found that it was not sufficient for handling the `aborts_if` condition (eg. [Line 21 in address_spec.move](https://github.com/MystenLabs/sui/pull/7149/files#diff-68dad675074fce7cf8afcdb47f72a47d68b7a11927b1f886dc8964443e57dceaR21)). As a result, I added a function body, which now makes it redundant.

How about removing the axioms and keeping only the function body? I am unsure if this will look weird or if implementing `aborts_if` condition seems unnecessary.

Reply:
@meng-xu-cs would be the best person to provide feedback here, but I will try.

I think that the idea of removing axioms and providing an explicit implementation should work. However, the current implementation in Boogie is different than the one in the runtime and I feel like we would need some way of convincing ourselves that they are equivalent. In other words, that two different byte vectors always produce two different addresses and that two same vectors never provide two different addresses. I am not sure if the current implementation conforms to this criterion. In my mind, the simplest (though I am not sure the easiest) way to proceed would be to have the same implementation in Boogie that we have in the runtime). This would be analogous to the model for `type_name::get` native function in core Move repo that has been implemented recently in https://github.com/move-language/move/pull/821

Reply:
thanks @awelc.
Do you think relying on axioms will be effective if we use the latest version of the move that includes those features - `type_name::get`?
Or is there anything else we should consider doing? I plan on testing it myself, but I wanted to get your thoughts first.